internal/venice-test-common/src/integrationTest/java/com/linkedin/venice/server/VeniceServerTest.java

```diff
@@ -182,6 +179,42 @@ public void testCheckBeforeJointClusterBeforeHelixInitializingCluster() throws E
     }
   }
 
+  @Test
+  public void testStartServerAndShutdownWithPartitionAssignmentVerification() {
+    try (VeniceClusterWrapper cluster = ServiceFactory.getVeniceCluster(1, 0, 0)) {
+      Properties featureProperties = new Properties();
+      featureProperties.setProperty(SERVER_ENABLE_SERVER_ALLOW_LIST, Boolean.toString(true));
+      featureProperties.setProperty(SERVER_IS_AUTO_JOIN, Boolean.toString(true));
+      cluster.addVeniceServer(featureProperties, new Properties());
+      VeniceServerWrapper server = cluster.getVeniceServers().get(0);
+      Assert.assertTrue(server.getVeniceServer().isStarted());
+      StorageService storageService = server.getVeniceServer().getStorageService();
+      StorageEngineRepository repository = storageService.getStorageEngineRepository();
+      Assert
+          .assertTrue(repository.getAllLocalStorageEngines().isEmpty(), "New node should not have any storage engine.");
+
+      // Create a storage engine.
+      String storeName = Version.composeKafkaTopic(cluster.createStore(1), 1);
+      Assert.assertEquals(repository.getAllLocalStorageEngines().size(), 1);
+      Assert.assertTrue(server.getVeniceServer().getHelixParticipationService().isRunning());
+      Assert.assertEquals(storageService.getStorageEngine(storeName).getPartitionIds().size(), 3);
+
+      cluster.stopVeniceServer(server.getPort());
+
+      // Create new servers so partition assignment is removed for the offline participant
+      cluster.addVeniceServer(featureProperties, new Properties());
+      cluster.addVeniceServer(featureProperties, new Properties());
+
+      cluster.restartVeniceServer(server.getPort());
+      repository = server.getVeniceServer().getStorageService().getStorageEngineRepository();
+      StorageEngineRepository storageEngineRepository = repository;
+      TestUtils.waitForNonDeterministicAssertion(
+          3,
+          TimeUnit.SECONDS,
+          () -> Assert.assertEquals(storageEngineRepository.getAllLocalStorageEngines().size(), 0));
```

Comment:
I think this whole section needs to be wrapped in `TestUtils.waitForNonDeterministicAssertion`. I'll let @ZacAttack confirm.

Reply:
Nah this is actually ok.  We're basically just doing the restart and waiting for what we need to happen.  In which case we need to only wait on the assert.